lib/Basic/Unix/TaskQueue.inc

```diff
@@ -320,164 +321,320 @@ void TaskQueue::addTask(const char *ExecPath, ArrayRef<const char *> Args,
-         !ExecutingTasks.empty()) {
-    // Enqueue additional tasks, if we have additional tasks, we aren't
-    // already at the parallel limit, and no earlier subtasks have failed.
-    while (!SubtaskFailed && !QueuedTasks.empty() &&
-           ExecutingTasks.size() < MaxNumberOfParallelTasks) {
-      std::unique_ptr<Task> T(QueuedTasks.front().release());
-      QueuedTasks.pop();
-      if (T->execute())
-        return true;
+public:
+  TaskShepherd(std::queue<std::unique_ptr<Task>> &TasksToExecute,
+               const unsigned NumberOfParallelTasks,
+               const TaskQueue::TaskBeganCallback BeganCallBack,
+               const TaskQueue::TaskFinishedCallback FinishedCallBack,
+               const TaskQueue::TaskSignalledCallback SignalledCallBack,
+               UnifiedStatsReporter *Stats)
+      : TasksToExecute(TasksToExecute),
+        MaxNumberOfParallelTasks(
+            NumberOfParallelTasks == 0 ? 1 : NumberOfParallelTasks),
+        BeganCallBack(BeganCallBack), FinishedCallBack(FinishedCallBack),
+        SignalledCallBack(SignalledCallBack), Stats(Stats) {}
+
+  /// Run the tasks in the queue, \return true on error.
+  bool executeTasks();
+
+private:
+  bool isFinishedExecutingTasks() const {
+    return ExecutingTasks.empty() && TasksToExecute.empty();
+  }
 
-      pid_t Pid = T->getPid();
+  /// Start up tasks if we aren't
+  /// already at the parallel limit, and no earlier subtasks have failed.
+  /// \return true for error;
+  bool startUpSomeTasks();
 
-      if (Began) {
-        Began(Pid, T->getContext());
-      }
+  /// Take ownership of the next task to start. Start it, and \return a
+  /// unique pointer if no error.
+  std::unique_ptr<Task> startNextTask();
 
-      PollFds.push_back({ T->getPipe(), POLLIN | POLLPRI | POLLHUP, 0 });
-      // We should also poll T->getErrorPipe(), but this introduces timing
-      // issues with shutting down the task after reading getPipe().
-      ExecutingTasks[Pid] = std::move(T);
-    }
+  // TODO: factor the polling routines and data structures into a separate class
 
-    assert(!PollFds.empty() &&
-           "We should only call poll() if we have fds to watch!");
-    int ReadyFdCount = poll(PollFds.data(), PollFds.size(), -1);
-    if (ReadyFdCount == -1) {
-      // Recover from error, if possible.
-      if (errno == EAGAIN || errno == EINTR)
-        continue;
+  /// Enter the task and its outputs in this TaskShepherd's data structures so
+  /// it can be polled.
+  void startPollingFdsOfTask(const Task *T);
+
+  void stopPolling(std::vector<int>);
+
+  enum class PollResult { hardError, softError, noError };
+  PollResult pollTheFds();
+
+  /// \return None on error
+  Optional<std::vector<int>> readFromReadyFdsReturningFinishedOnes();
+
+  /// \return true if fd "hung up"
+  bool reactToFdEvents(short events, int fd, Task &T);
+};
+
+/// Do everything required to handle a task when its output has hung up.
+class TaskCompleter {
```

Comment:
Why bother making this a separate class?

Reply:
I added a comment:
`/// Encapsulates the knowledge of how to wait, and then how to clean up.`
I like that encapsulation, but what are you thinking?